modules.go

```diff
@@ -0,0 +1,455 @@
+package goja
+
+import (
+	"errors"
+)
+
+type HostResolveImportedModuleFunc func(referencingScriptOrModule interface{}, specifier string) (ModuleRecord, error)
+
+// TODO most things here probably should be unexported and names should be revised before merged in master
+// Record should probably be dropped from everywhere
+
+// ModuleRecord is the common interface for module record as defined in the EcmaScript specification
+type ModuleRecord interface {
+	GetExportedNames(resolveset ...ModuleRecord) []string
+	ResolveExport(exportName string, resolveset ...ResolveSetElement) (*ResolvedBinding, bool)
+	Link() error
+	Evaluate(*Runtime) *Promise
+}
+
+type CyclicModuleRecordStatus uint8
```

Comment:
The current naming scheme is "whatever it was called in the specification, or something very similar when it's around it". Which seems to be somewhat what you have been doing and given that this makes the development *a lot* easier, I do like it. 

But on that note while I did add ModuleInstance (and CyclicModuleInstance) as types in addition to everything in the specification, I do think we do need .... uh more wrappers :(.

I was thinking of something similar to what you did with [DynamicArray](https://pkg.go.dev/github.com/dop251/goja#DynamicArray) - the user of the API provides some implementation but the actual final object/type for the (Cylic)ModuleRecord(Instance) is something wrapping the user provided data.

This will IMO make the `modules.go` a lot more understandable and will help with the fact that currently any `CyclicModuleRecord` implementation should know to run [CyclicModuleRecordEvaluate](https://github.com/mstoykov/goja/blob/5f81441173723ebb181873312ffaa98604cdd2b0/modules.go#L149-L176) as it's implementation of `Evaluate`. 

I did try this before doing the async module evaluation support, but it turned out I was adding types everywhere and decided it's not worth it. But with the current amount of things in the [evaluationState](https://github.com/mstoykov/goja/blob/modules/modules.go#L123-L133) it seems like it will be really nice idea. This also will mean that the actual [RecordStatus](https://github.com/mstoykov/goja/blob/5f81441173723ebb181873312ffaa98604cdd2b0/modules.go#L22-L29) will be part of the ModuleRecord. 

(Note: earlier version had the state as part of the `CyclicModuleRecord` interface - but that meant every module needed to implement the same Get/Set methods >.> which was quite cumbersome even with a helper type you can embed to do it for you.)

I am going to work on this next as long as you don't come back with better idea or some other feedback.